### PR TITLE
refactor(bundler): New Gemfile parser based on "tree-sitter"

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

This PR introduces "tree-sitter" for parsing dependencies from complex languages.

## Context

The current regex-based implementation is hard to maintain.
It blocks implementation of #4789, which should be straightforward with this new parsing approach.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
Or, another possibility that it has something to do with our aggressive CI caching

**Review by viceice — COMMENTED** [review]
another parser? isn't your self written parser good enough?

it's probably related to node v18 issues?

**Comment by zharinov** [comment]
Nah, it's not good enough 😂

**Comment by zharinov** [comment]
Starting from `0.21.0`, the NAPI-based prebuilds are used, which seems to work:

image

The problem is some (e.g. `ruby` and `scala`) are not updated yet.

File: lib/modules/manager/bundler/extract.spec.ts [hunk 1]
```diff
@@ -135,10 +135,10 @@ describe('modules/manager/bundler/extract', () => {
       'Gemfile',
     );
     expect(res?.deps).toMatchObject([
```
Comment (viceice) [inline lib/modules/manager/bundler/extract.spec.ts:137]: why the order changes?

File: lib/modules/manager/bundler/parser.ts [hunk 2]
```diff
@@ -0,0 +1,327 @@
+import Parser, { Query, SyntaxNode } from 'tree-sitter';
```
Comment (viceice) [inline lib/modules/manager/bundler/parser.ts:1]: it seems it's something like antlr?

does it make sense to use it for more languages?

File: lib/modules/manager/bundler/parser.ts [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        const source = extractRegistry(node);
+        if (source) {
+          globalSources.unshift(source);
+        }
+      }
+    }
+  }
+
+  const cachedGroups: WeakMap<SyntaxNode, string[]> = new WeakMap();
+  const cachedSources: WeakMap<SyntaxNode, string> = new WeakMap();
```
Comment (viceice) [inline lib/modules/manager/bundler/parser.ts:196]:
```suggestion
  const cachedGroups = new WeakMap<SyntaxNode, string[]>();
  const cachedSources = new WeakMap<SyntaxNode, string>();
```
😉

**Review by viceice — COMMENTED** [review]
looks complicated 🫣

File: lib/modules/manager/bundler/parser.ts (hunk 2 above)
Comment (zharinov) [reply lib/modules/manager/bundler/parser.ts:1]: Yes, it is similar to Antlr, now it's gaining popularity among text editor enthusiasts (specifically, Neovim), and it powers GitHub's syntax highlight.

**Pull request closed without merging**